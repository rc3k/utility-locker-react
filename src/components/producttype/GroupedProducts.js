import React from 'react';

import ProductListItem from '../product/ProductListItem';
import SortBySelect from '../SortBySelect';

export default ({
  item, products, productIds, loadProducts, typeParams,
}) => (
  <div>
    {productIds[item.key].length > 1 ? (
      <div className="row">
        <div className="col-md-6 offset-md-6">
          <SortBySelect
            column={typeParams[item.key].column}
            direction={typeParams[item.key].direction}
            setSort={(column, direction) => loadProducts({ column, direction, type: item.key })}
          />
        </div>
      </div>
    ) : null}
    <div className="p-4">
      {productIds[item.key].map((itemId) => (
        <ProductListItem
          key={itemId}
          item={products[itemId]}
        />
      ))}
    </div>
  </div>
);
