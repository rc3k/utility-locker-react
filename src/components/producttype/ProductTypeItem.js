import React from 'react';
import { CSSTransition } from 'react-transition-group';

import GroupedProducts from './GroupedProducts';

export default ({
  item, toggleAccordionItemOpen, isOpen, products, productIds, loadProducts, typeParams,
}) => (
  <div className="product-type-item card mb-3">
    <button
      type="button"
      onClick={() => toggleAccordionItemOpen({
        name: 'productTypes',
        itemId: item.key,
      })}
      className="accordion-toggle"
    >
      <div
        className="card-header justify-content-between d-flex"
      >
        <span>{item.value}</span>
        <span>{!isOpen ? '+' : '-'}</span>
      </div>
    </button>
    <CSSTransition
      in={isOpen}
      timeout={200}
      classNames="product-transition"
    >
      <div className="container">
        {item.key in productIds && isOpen
          ? (
            <GroupedProducts
              item={item}
              products={products}
              productIds={productIds}
              loadProducts={loadProducts}
              typeParams={typeParams}
            />
          )
          : <div />}
      </div>
    </CSSTransition>
  </div>
);
