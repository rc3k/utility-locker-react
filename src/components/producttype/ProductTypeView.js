import React from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import { productTypes as productTypesSlice } from '../../slices/productTypes';
import * as thunks from '../../slices/thunks';
import { ui } from '../../slices/ui';
import Preloader from '../Preloader';
import ProductTypeItem from './ProductTypeItem';
import Header from '../Header';

export class ProductTypeView extends React.Component {
  componentDidMount() {
    const { loadProductTypes } = this.props;
    loadProductTypes();
  }

  render() {
    const {
      products, productTypes, openTypes, history, toggleAccordionAndLoad, loadProductsByType,
    } = this.props;
    const { items, loading, typeParams } = productTypes;

    if (loading) {
      return (
        <Preloader />
      )
    }

    return (
      <div className="list-view">
        <Header
          heading="Product types"
          sortParams={products.params}
          groupByField="type"
          setGroupBy={(groupBy) => history.push(`/${groupBy}`)}
        />
        <div className="list-group">
          {items.allIds.map((itemId) => (
            <ProductTypeItem
              key={itemId}
              item={items.byId[itemId]}
              isOpen={openTypes.includes(itemId)}
              productIds={items.productIdsByType}
              products={products.items.byId}
              typeParams={typeParams}
              toggleAccordionItemOpen={toggleAccordionAndLoad}
              loadProducts={(params) => loadProductsByType(params)}
            />
          ))}
        </div>
      </div>
    );
  }
}

export const TypeViewContainer = withRouter(
  connect((state) => ({
    products: state.products,
    productTypes: state.productTypes,
    openTypes: state.ui.accordion.productTypes,
  }), { ...productTypesSlice.actions, ...ui.actions, ...thunks })(ProductTypeView),
);
