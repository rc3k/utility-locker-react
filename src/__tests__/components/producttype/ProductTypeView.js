import React from 'react';
import renderer from 'react-test-renderer';

import { BrowserRouter as Router } from 'react-router-dom';
import { ProductTypeView } from '../../../components/producttype/ProductTypeView';

it('renders correctly when data is loading', () => {
  const listView = renderer.create(
    <Router>
      <ProductTypeView
        productTypes={{loading: true}}
        loadProductTypes={() => {}}
      />
    </Router>,
  ).toJSON();
  expect(listView).toMatchSnapshot();
});

it('renders correctly when items have loaded', () => {

  const listView = renderer.create(
    <Router>
      <ProductTypeView
        productTypes={{
          loading: false,
          items: {
            allIds: ['book', 'electrical'],
            byId: {
              ['book']: {
                key: 'book',
                value: 'Book',
              },
              ['electrical']: {
                key: 'electrical',
                value: 'Electrical',
              }
            },
            productIdsByType: {
              book: ['123-X'],
              electrical: ['123-Y']
            }
          }
        }}
        products={
          {
            loading: false,
            items: {
              byId: {
                '123-X': {
                  id: '123-X',
                  name: 'Item one',
                  description: 'Item description',
                  weight: 123,
                  price: {
                    value: 156.98
                  },
                  type: 'book',
                  type_name: 'Book',
                },
                '124-Y': {
                  id: '123-Y',
                  name: 'Item two',
                  description: 'Item description 2',
                  weight: 987,
                  price: {
                    value: 1.23
                  },
                  type: 'electrical',
                  type_name: 'Electrical',
                },
              },
              allIds: ['123-X', '124-Y',],
            },
            params: {
              column: 'name',
              direction: 'asc'
            },
          }
        }
        loadProducts={() => {}}
        loadProductTypes={() => {}}
        openTypes={[]}
      />
    </Router>,
  ).toJSON();
  expect(listView).toMatchSnapshot();
});
