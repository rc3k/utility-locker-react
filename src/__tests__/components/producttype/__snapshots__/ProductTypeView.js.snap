// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when data is loading 1`] = `
<div
  className="loader"
>
  Loading...
</div>
`;

exports[`renders correctly when items have loaded 1`] = `
<div
  className="list-view"
>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <h1>
        Product types
      </h1>
    </div>
    <div
      className="col-md-6"
    >
      <div
        className="d-flex w-100 form-group p-2"
      >
        <label
          className="pt-2 pr-1 text-right"
          htmlFor="group-by-select"
        >
          Group by:
        </label>
        <select
          className="form-control"
          id="group-by-select"
          onChange={[Function]}
          value="type"
        >
          <option
            value=""
          >
            None
          </option>
          <option
            value="type"
          >
            Type
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    className="list-group"
  >
    <div
      className="product-type-item card mb-3"
    >
      <button
        className="accordion-toggle"
        onClick={[Function]}
        type="button"
      >
        <div
          className="card-header justify-content-between d-flex"
        >
          <span>
            Book
          </span>
          <span>
            +
          </span>
        </div>
      </button>
      <div
        className="container"
      >
        <div />
      </div>
    </div>
    <div
      className="product-type-item card mb-3"
    >
      <button
        className="accordion-toggle"
        onClick={[Function]}
        type="button"
      >
        <div
          className="card-header justify-content-between d-flex"
        >
          <span>
            Electrical
          </span>
          <span>
            +
          </span>
        </div>
      </button>
      <div
        className="container"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;
